### PR TITLE
Reduce regressor peak memory by chunking translate_probs_across_borders

### PR DESCRIPTION
**Pull request by LeoGrin**
## Summary

- `_cdf` inside `translate_probs_across_borders` allocates several `(n_test, num_buckets)` fp32 intermediates. At large `n_test` this dominates peak memory in `TabPFNRegressor.predict`, far beyond the size of the returned array.
- The transform is row-independent, so we chunk the leading dimension when the flattened batch would produce intermediates larger than a fixed element budget. Output is numerically identical to the unchunked path; small batches keep the original fast path.
- PR #745 handled the `n_estimators` axis (on-the-fly accumulation); this handles the `n_test` axis. Complementary fixes.

Measured on CUDA (H100, `n_train=10k`, `features=50`, `fit_mode=fit_preprocessors`, `n_estimators=1`):

| n_test   | orig peak | patched peak |
|---------:|----------:|-------------:|
| 32_000   |   7.46 GB |      4.55 GB |
| 100_000  |  23.09 GB |     11.47 GB |
| 250_000  |  57.61 GB |     22.80 GB |

Predict wall time changes by <1% across 5 trials at each size, and is slightly faster at the largest sizes (reduced allocator pressure).

## Test plan

- [x] New parametrized test in `tests/test_utils.py` asserting the chunked path matches the unchunked reference bit-for-bit (`atol=0, rtol=0`) for batch sizes in {1, 32, 4097}.
- [x] `pytest tests/test_utils.py` — 13 passed.
- [x] Smoke predict with `n_estimators=2` returns expected shape/values.

## Note to reviewer

Rename `changelog/PR.fixed.md` → `changelog/<this-PR-number>.fixed.md` once the PR number is assigned.

---

> [!NOTE]
> **Medium Risk**
> Touches core tensor math used during prediction and introduces a chunking loop that could affect performance or shape handling on edge cases, though tests assert bit-identical outputs.
> 
> **Overview**
> Reduces `TabPFNRegressor.predict` peak GPU memory for very large test sets by chunking `translate_probs_across_borders` over flattened batch rows, keeping `_cdf` intermediates bounded and avoiding `torch.cat` transients via preallocated output.
> 
> Adds `_translate_probs_across_borders_unchunked` as a reference/fast path and new tests asserting bit-exact equality with the unchunked implementation and verifying the chunked dispatch across arbitrary batch shapes; includes a changelog entry documenting the memory reduction.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit c939459f9ebe2854f602da2fe39c70a9d00f318d. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: src/tabpfn/utils.py [hunk 1]
```diff
@@ -345,6 +364,13 @@ def translate_probs_across_borders(
...
 
+    For large batches the computation is chunked so that the peak memory
+    footprint of the intermediate ``(batch, len(to))`` tensors allocated
+    inside ``_cdf`` stays bounded. All batch dimensions are flattened
+    before chunking, so the memory cap holds regardless of which batch
+    dimension is large (e.g. `(n_estimators, n_test, num_buckets)`). The
+    output is numerically identical to the unchunked version.
+
     Args:
         logits: The logits defining the distribution to translate.
```
Comment (bejaeger) [inline src/tabpfn/utils.py]: nit: could we document the possible shape of the logits?

File: src/tabpfn/utils.py [hunk 2]
```diff
@@ -337,6 +337,25 @@ def _cdf(logits: torch.Tensor, borders: torch.Tensor, ys: torch.Tensor) -> torch
     return prob_left_of_ys.clip(0.0, 1.0)
 
 
+def _translate_probs_across_borders_unchunked(
+    logits: torch.Tensor,
+    *,
+    frm: torch.Tensor,
```
Comment (bejaeger) [inline src/tabpfn/utils.py:343]:
```suggestion
    from: torch.Tensor,
```

File: src/tabpfn/utils.py [hunk 3]
```diff
@@ -353,11 +379,33 @@ def translate_probs_across_borders(
     Returns:
         The translated probabilities.
     """
-    prob_left = _cdf(logits, borders=frm, ys=to)
-    prob_left[..., 0] = 0.0
-    prob_left[..., -1] = 1.0
-
-    return (prob_left[..., 1:] - prob_left[..., :-1]).clamp_min(0.0)
+    batch_shape = logits.shape[:-1]
+    num_buckets_frm = logits.shape[-1]
```
Comment (bejaeger) [inline src/tabpfn/utils.py:392]:
```suggestion
    num_buckets_from = logits.shape[-1]
```

File: src/tabpfn/utils.py [hunk 4]
```diff
@@ -337,6 +337,25 @@ def _cdf(logits: torch.Tensor, borders: torch.Tensor, ys: torch.Tensor) -> torch
...
+    prob_left[..., 0] = 0.0
+    prob_left[..., -1] = 1.0
+    return (prob_left[..., 1:] - prob_left[..., :-1]).clamp_min(0.0)
+
+
+# `_cdf` allocates ~8 intermediate tensors of shape (batch, len(to)). Targeting
+# `chunk_size * len(to) <= _TRANSLATE_CHUNK_BUDGET_ELEMENTS` keeps each transient
+# around ~80 MB (fp32) and the total under ~1 GB, which holds translate_probs's
+# contribution to peak memory roughly constant in n_test.
+_TRANSLATE_CHUNK_BUDGET_ELEMENTS = 20_000_000
```
Comment (bejaeger) [inline src/tabpfn/utils.py:356]: nit: could also pass this as default to the FN; then we don't need to mock it in the tests.

**Review by bejaeger — APPROVED** [review]
LGTM! just some nits

File: src/tabpfn/utils.py (hunk 2 above)
Comment (LeoGrin) [reply src/tabpfn/utils.py:343]: Can't apply this one — `from` is a reserved keyword in Python, so it can't be used as a parameter name. The existing code uses `frm` to sidestep that. Happy to rename to `from_` if you prefer that spelling; otherwise leaving as-is.

File: src/tabpfn/utils.py (hunk 3 above)
Comment (LeoGrin) [reply src/tabpfn/utils.py:392]: This one is tied to the `from` rename above — since we can't use `from` as a Python identifier, keeping `num_buckets_frm` for consistency with the existing `frm` kwarg.

Comment (bejaeger) [reply src/tabpfn/utils.py:392]: Oh..okay! fine with me then :)
